SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/RemotePossessionComponent.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/RemotePossessionComponent.h"
+#include "LoadBalancing/AbstractLBStrategy.h"
+
+#include "Engine/World.h"
+
+DEFINE_LOG_CATEGORY(LogRemotePossessionComponent);
+
+URemotePossessionComponent::URemotePossessionComponent(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+	, PendingDestroy(false)
+{
+	PrimaryComponentTick.bCanEverTick = true;
+	PrimaryComponentTick.bStartWithTickEnabled = true;
+
+#if ENGINE_MINOR_VERSION <= 23
+	bReplicates = true;
+#else
+	SetIsReplicatedByDefault(true);
+#endif
+}
+
+void URemotePossessionComponent::OnAuthorityGained()
+{
+	AController* Controller = Cast<AController>(GetOwner());
+	ensure(Controller);
+	if (Target == nullptr)
+	{
+		UE_LOG(LogRemotePossessionComponent, Error, TEXT("Target is null"));
+		return;
+	}
+	if (!Target->HasAuthority())
+	{
+		UE_LOG(LogRemotePossessionComponent, Verbose, TEXT("Worker is not authoritative over target: %s"), *Target->GetName());
+		return;
+	}
+	else
+	{
+		if (EvaluatePossess())
+		{
+			UE_LOG(LogRemotePossessionComponent, Log, TEXT("Possess(%s)"), *Target->GetName());
+			Controller->Possess(Target);
+		}
+		else
+		{
+			UE_LOG(LogRemotePossessionComponent, Log, TEXT("EvaluatePossess(%s) failed"), *Target->GetName());
+		}
+		MarkToDestroy();
+	}
+}
+
+void URemotePossessionComponent::TickComponent(float DeltaTime, enum ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
+	if (PendingDestroy)
+	{
+		UE_LOG(LogRemotePossessionComponent, Log, TEXT("DestroyComponent"));
+		DestroyComponent();
+	}
+}
+
+bool URemotePossessionComponent::EvaluatePossess()
+{
+	if (GetClass()->HasAnyClassFlags(CLASS_CompiledFromBlueprint) || !GetClass()->HasAnyClassFlags(CLASS_Native))
```

Comment:
This looks like it what BlueprintNativeEvent does. I do not think there is a need to reivent it?

Reply:
If without this check, our code will looks like:
bool URemotePossessionComponent::EvaluatePossess()
{
	return ReceiveEvaluatePossess();
}
The default return value of ReceiveEvaluatePossess is false. I have no idea to change it to true. It means the customers should have their inherited blueprint RemotePossessionComponent.

Reply:
What I'm saying is I think we do not need ReceiveEvaluatePossess at all.
EvaluatePossess would be a BlueprintNativeEvent, which can be overrideon both from blueprints or C++.
This is done through the automatic creation by the UnrealHeaderTool of a method EvaluatePossess_Implementation.
This method is virtual. And the implementation of EvaluatePossess does the job of selecting which implementation to call, blueprint or C++.

Reply:
Good idea, I had updated it now.